File: src/Application Framework/Structures/DataGridRowHeightSetter.vb
Imports System
Imports System.Reflection

Public Class DataGridRowHeightSetter

    Private dg As DataGrid
    Private rowObjects As ArrayList

    Public Sub New(ByVal dg As DataGrid)
        Me.dg = dg
        InitHeights()
    End Sub

    Default Public Property Item(ByVal row As Integer) As Integer
        Get
            Try
                Dim pi As PropertyInfo = rowObjects(row).GetType().GetProperty("Height")
                Return Fix(pi.GetValue(rowObjects(row), Nothing))
            Catch
                Throw New ArgumentException("invalid row index")
            End Try
        End Get
        Set(ByVal Value As Integer)
            Try
                Dim pi As PropertyInfo = rowObjects(row).GetType().GetProperty("Height")
                pi.SetValue(rowObjects(row), Value, Nothing)
            Catch
                Throw New ArgumentException("Invalid row index")
            End Try
        End Set
    End Property

    Private Sub InitHeights()
        Dim mi As MethodInfo = dg.GetType().GetMethod("get_DataGridRows", BindingFlags.FlattenHierarchy Or BindingFlags.IgnoreCase Or BindingFlags.Instance Or BindingFlags.NonPublic Or BindingFlags.Public Or BindingFlags.Static)

        Dim dgra As System.Array = CType(mi.Invoke(Me.dg, Nothing), System.Array)

        rowObjects = New ArrayList
        Dim dgrr As Object
        For Each dgrr In dgra
            If dgrr.ToString().EndsWith("DataGridRelationshipRow") = True Then
                rowObjects.Add(dgrr)
            End If
        Next dgrr
    End Sub

End Class

File: src/Application Framework/Structures/DataGridRowHeightSetter.cs
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using System;
using System.Collections;
using System.Reflection;
using System.Windows.Forms;

namespace FSM
{
    public class DataGridRowHeightSetter
    {
        private DataGrid dg;
        private ArrayList rowObjects;

        public DataGridRowHeightSetter(DataGrid dg)
        {
            this.dg = dg;
            InitHeights();
        }

        public int this[int row]
        {
            get
            {
                try
                {
                    var pi = rowObjects[row].GetType().GetProperty("Height");
                    return Conversions.ToInteger(Conversion.Fix(pi.GetValue(rowObjects[row], null)));
                }
                catch
                {
                    throw new ArgumentException("invalid row index");
                }
            }

            set
            {
                try
                {
                    var pi = rowObjects[row].GetType().GetProperty("Height");
                    pi.SetValue(rowObjects[row], value, null);
                }
                catch
                {
                    throw new ArgumentException("Invalid row index");
                }
            }
        }

        private void InitHeights()
        {
            var mi = dg.GetType().GetMethod("get_DataGridRows", BindingFlags.FlattenHierarchy | BindingFlags.IgnoreCase | BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static);
            Array dgra = (Array)mi.Invoke(dg, null);
            rowObjects = new ArrayList();
            object dgrr;
            foreach (var dgrr in dgra)
            {
                if (dgrr.ToString().EndsWith("DataGridRelationshipRow") == true)
                {
                    rowObjects.Add(dgrr);
                }
            }
        }
    }
}